Simplify imports by reexporting core in api

javajawa opened the pull request:
This merge performs a number of structural refactors:

By exporting the `InputEvent` and `OutputEvent` base classes in `mewbot.api.v1`, the IO libraries no longer need to import from both that and the `mewbot.core` module. This makes the import diagram simpler, and highlights the difference between the core and the plug and play API.

The restrictions on API versions have been relaxed or removed, with tools like `mewbot.loader` testing against the underlying interfaces to confirm API compatability.

javajawa committed:
Simplify imports by reexporting core in api

By exporting the InputEvent and OutputEvent base classes in
`mewbot.api.v1`, the IO libraries no longer need to import from
both that and the `mewbot.core` module. This makes the import
diagram simpler, and highlights the difference between the core
and the plug and play API.

javajawa committed:
Remove most API version checks

Going forward, most of the API compatibilty will be defined by the
protocols in mewbot.core. Some of these checks were already disabled
for Behaviours to work at all.

javajawa committed:
Validate loaded classes against the core interfaces

When loading using `mewbot.loader`, we do not need to require that
all classes are using the registry system, only that they implement
the correct interfaces. This change detatches the loader from the
component module, and also allows future work to use different base
class systems.

This change also means that loaded files are checked that they are
the correct interface based on the `kind:` field value.

exachixkitsune commented on src/mewbot/core.py:
```diff
@@ -151,8 +118,51 @@ async def process(self, event: InputEvent) -> None:
+        raise ValueError(f"Invalid value {value}")
+
+
 __all__ = [
```
Should these have a particular ordering, i.e. same as in file for findability?
 I cannot fathom the suggestions interface so you'll have to give me a bit for a suggestion.

javajawa replied:
I'm not sure it matters, really? The order in the file is to prevent forward definition issues, the order here was mostly grouped by purpose.
There's no guarantee that the things being exported are even first defined in this file (part of this PR is in fact exporting things from another module)

ajCameron replied:
I think, in the general case, re-ordering to increase clarity is a good thing. Not sure how to do it here though. But we should be mindful of that in future - I've seen people draw deep - largely wrong - inferences from structure.

exachixkitsune replied:
I think it could just be left as developer preference and is far and away not a thing we should worry about. I'd put it in the level of minor readability.

ajCameron replied:
Makes sense! We can always change it later on a readability pass.

exachixkitsune reviewed (dismissed):
I don't see much in the way of issues here, only a convention thingamabob on __all_- but that's a minor thought

ajCameron commented on src/mewbot/loader.py:
```diff
@@ -103,11 +96,25 @@ def load_component(config: ConfigBlock) -> Component:
+    # Verify that the implementation class matches the interface we got from
+    # the `kind:` hint.
+    if not issubclass(target_class, interface):
         raise Exception("Trying to load a non-registered class from config")
```
This will be a theme with me when it comes to error messages. I'm a fan of maximal detail to give the user a debugging headstart. Though this does result in messier code.

Suggestion: Include a list of the currently registered classes - ideally broken down by type - so the user can easily see if they've just typod or if there is a more fundamental problem and the object they're trying to specify hasn't been loaded at all.

ajCameron commented on src/mewbot/loader.py:
```diff
@@ -103,11 +96,25 @@ def load_component(config: ConfigBlock) -> Component:
     component = target_class.__call__(uid=config["uuid"], **config["properties"])
 
-    assert isinstance(component, Component)
+    # Verify the instance implements a valid interface.
```
Suggest some kind of error message here for if the assert fails.

javajawa replied:
I'm still not over the fact that python's assert doesn't have `assert([bool], [str])`

ajCameron replied:
...

That would be really kinda neat.

I mean, they'll probably reject it, but I'm weirdly game for drafting a PEP

javajawa replied:
It turns out, it does work like that! It's just not obvious because `assert` _doesn't use brackets_.

```python
>>> assert(False, "Hello")
>>> assert False, "Hello"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: Hello
```

The first one gets interpreted as `Tuple[Bool=False, Str="Hello"]` which is truthy.

(Thanks, I hate it. Also I thought they would have fixed that the same time as they fixed `print`)

ajCameron replied:
Ahhhhhhhhh!

Yes. I, too, am far from a fan!

javajawa committed:
Update exceptions in mewbot.loader

 - Assertions now have explanatory error messages
 - The use of `Exception` has been replaced with an appropriate error

ajCameron reviewed (approved):
Sensible organisation and structural changes!

(Just did another pass through, and nothing jumps out to suggest - all looks good).
